### PR TITLE
Add test-runner dependencies in build def

### PR DESCRIPTION
**Pull request by samwestmoreland**
Add test-runner dependencies in build def rather than in the pex tool

### CODE REVIEW

File: tools/wheel_resolver/test/BUILD [hunk 1]
```diff
@@ -6,8 +6,8 @@ python_test(
 
 gentest(
     name = "e2e_test",
-    data = ["//tools/wheel_resolver"],
+    data = ["//tools/wheel_resolver/src:resolve"],
     no_test_output = True,
-    test_cmd = "PKG=tensorflow; OUTS=$PKG.whl; export OUTS; $DATA --package \"$PKG\"; \
+    test_cmd = "PKG=cryptography; OUTS=$PKG.whl; export OUTS; $DATA --package \"$PKG\"; \
```
Comment (samwestmoreland) [inline tools/wheel_resolver/test/BUILD:11]: Wheel resolver tool now fails to download tensorflow. It also fails to download a subset of the things in third_party/python - I'll raise a separate issue for this.

File: build_defs/python.build_defs [hunk 2]
```diff
@@ -47,6 +47,10 @@ def python_library(name:str, srcs:list=[], resources:list=[], deps:list=[], visi
         labels += ['py:zip-unsafe']
     if srcs or resources:
         cmd = '$TOOLS_JARCAT z -d -o ${OUTS} -i .'
+
+        # For some reason jarcat isn't unflattening .whl files so do it manually here till I figure out why
+        unzip_cmd = f'for i in $(find . -name \'*.whl\'); do mv $i .; unzip $(basename "$i"); done'
```
Comment (peterebden) [inline build_defs/python.build_defs:52]: would be nice to figure this out, this will be a significant performance disadvantage if we have to recompress a bunch of files.

File: tools/jarcat/BUILD [hunk 3]
```diff
@@ -0,0 +1,6 @@
+filegroup(
```
Comment (peterebden) [inline tools/jarcat/BUILD:1]: shouldn't these tools just be the defaults, then plz will download them for us?

File: tools/please_pex/BUILD [hunk 4]
```diff
@@ -1,7 +1,69 @@
-# TODO(jpoole): This should be a remove file on the plugin repo's release page and we should remove //_please:please_pex
-filegroup(
+subinclude("//build_defs:python")
```
Comment (peterebden) [inline tools/please_pex/BUILD:1]: Would suggest removing the top level `tools` dir, I'm not sure it adds much now it's in its own repo

File: tools/please_pex/BUILD [hunk 5]
```diff
@@ -1,7 +1,69 @@
...
+    name = "pex_main",
+    srcs = ["pex_main.go"],
+    deps = [
+        "//third_party/go:cli-init",
+        "//third_party/go:go-logging.v1",
+        "//tools/please_pex/pex",
+    ],
+)
+
+# Combine all the Python dependencies into one zipfile and concatenate them
```
Comment (peterebden) [inline tools/please_pex/BUILD:13]: Is this stuff still relevant? Thought maybe we were just going to make them normal dependencies?

File: .plzconfig [hunk 6]
```diff
@@ -7,31 +7,38 @@ PreloadBuildDefs = test/build_defs/test.build_defs
...
 [PluginConfig "test_runner"]
 ConfigKey = TestRunner
 DefaultValue = unittest
+Optional = true
 
-[PluginConfig "test_runner_bootstrap"]
-ConfigKey = TestRunnerBootstrap
+[PluginConfig "testrunner_deps"]
+ConfigKey = TestrunnerDeps
+DefaultValue = @self//third_party/python:unittest_bootstrap
```
Comment (samwestmoreland) [inline .plzconfig:41]: This should probably be inferred from the TestRunner field

File: build_defs/python.build_defs (hunk 2 above)
Comment (samwestmoreland) [reply build_defs/python.build_defs]: Ah I was missing a `-s .whl` in the jarcat command

File: tools/please_pex/BUILD [hunk 7]
```diff
@@ -1,7 +1,40 @@
...
+    ],
+    outs = ["please_pex"],
+    binary = True,
+    cmd = [
+        '"$TOOL" z -d -i . -o "$OUTS" --preamble_file $(location :pex_main)',
+    ],
+    tools = ["//tools:arcat"],
     visibility = ["PUBLIC"],
+    labels = ["hlink:plz-out/package"],
+)
```
Comment (tiagovtristao) [inline tools/please_pex/BUILD:26]: I don't think we need this rule anymore. We aren't zipping up the bootstrap deps to the pex tool anymore. I reckon the rule above and this one can be merge, without needing arcat?

File: tools/please_pex/pex/pex_test_main.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            _original_xml_file(self, fr, analysis, *args, **kvargs)
+    coverage_control.XmlReporter.xml_file = _xml_file
+    return coverage
+
+
+def main():
+    """Runs the tests. Returns an appropriate exit code."""
+    args = [arg for arg in sys.argv[1:]]
+    # Add .bootstrap dir to path, after the initial pex entry
+    sys.path = sys.path[:1] + [os.path.join(sys.path[0], '.bootstrap')] + sys.path[1:]
```
Comment (tiagovtristao) [inline tools/please_pex/pex/pex_test_main.py:35]: I think this isn't needed anymore, given that the `.bootstrap` dir shouldn't exist?

File: tools/please_pex/pex/pex_test_main.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    if os.getenv('COVERAGE'):
+        # It's important that we run coverage while we load the tests otherwise
+        # we get no coverage for import statements etc.
+        cov = initialise_coverage().coverage(data_file=None)
+        cov.exclude(r'^import\b')
+        cov.exclude(r'^from\b')
+        cov.start()
+        result = run_tests(args)
+        cov.stop()
+        omissions = ['*/third_party/*', '*/.bootstrap/*']
```
Comment (tiagovtristao) [inline tools/please_pex/pex/pex_test_main.py:45]: The bootstrap entry might can likely be removed if .bootstrap dir is gone.

File: tools/please_pex/pex/pex.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+func (pw *Writer) SetDebugger(debugger Debugger) {
+	pw.pexStamp = "debug"
+
+	switch debugger {
+	case "pdb":
+		pw.debugger = filepath.Join(debuggersDir, "pdb.py")
+	case "debugpy":
+		pw.debugger = filepath.Join(debuggersDir, "debugpy.py")
+		pw.includeLibs = append(pw.includeLibs, ".bootstrap/debugpy")
```
Comment (tiagovtristao) [inline tools/please_pex/pex/pex.go:110]: I think this isn't needed anymore, if debugpy now lives in third_party?

File: tools/please_pex/pex/pex_main.py [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+        self.suffixes_by_length = sorted(self.suffixes, key=lambda x: -len(x))
+        # Identify all the possible modules we could handle.
+        self.modules = {}
+        if is_zipfile(sys.argv[0]):
+            zf = ZipFileWithPermissions(sys.argv[0])
+            for name in zf.namelist():
+                path, _ = self.splitext(name)
+                if path:
+                    if path.startswith('.bootstrap/'):
+                        path = path[len('.bootstrap/'):]
```
Comment (tiagovtristao) [inline tools/please_pex/pex/pex_main.py:131]: You might be able to do some cleanup around this init function too

**Review by tiagovtristao — COMMENTED** [review]
Looks good! Mostly clean up comments.

File: build_defs/python.build_defs [hunk 12]
```diff
@@ -305,7 +328,7 @@ def python_test(name:str, srcs:list, data:list|dict=[], resources:list=[], deps:
         data=data | {'_srcs': srcs} if isinstance(data, dict) else data + srcs,
         outs=[f'{name}.pex'],
         labels=labels + ['test_results_dir'],
-        cmd='$TOOL z -i . -s .pex.zip -s .whl --preamble_from="$SRC" --include_other --add_init_py --strict',
+        cmd=f'mv $SRC __main__.py && $TOOL z -i . -s .pex.zip -s .whl --preamble=\"{shebang}\" --include_other --add_init_py --strict',
```
[the remark this answers is not in the document]
Comment (samwestmoreland) [reply build_defs/python.build_defs:331]: Nah, don't think I needed to escape here. Thanks

File: third_party/python/BUILD [hunk 13]
```diff
@@ -56,18 +59,15 @@ python_wheel(
     version = "3.10.0.2",
 )
 
-python_wheel(
+pip_library(
```
[the remark this answers is not in the document]
Comment (samwestmoreland) [reply third_party/python/BUILD:62]: A lot of them had to be changed to pip_library to avoid circular dependencies with the wheel_resolver tool, but actually this one was fine as a python_wheel

File: tools/please_pex/BUILD (hunk 7 above)
Comment (samwestmoreland) [reply tools/please_pex/BUILD:26]: Good point!

File: tools/please_pex/pex/pex.go (hunk 10 above)
Comment (samwestmoreland) [reply tools/please_pex/pex/pex.go]: Actually `includeLibs` doesn't really do anything anymore... I'm wondering if debugpy is actually being pulled in anywhere now

File: build_defs/python.build_defs [hunk 14]
```diff
@@ -129,13 +130,16 @@ def python_binary(name:str, main:str, srcs:list=[], resources:list=[], out:str=N
...
-    shebang = shebang or interpreter or CONFIG.PYTHON.DEFAULT_INTERPRETER
+    interpreter = interpreter or CONFIG.PYTHON.DEFAULT_INTERPRETER
+    shebang = shebang or _set_shebang(interpreter, CONFIG.PYTHON.INTERPRETER_OPTIONS, site)
     zipsafe_flag = '' if zip_safe is False else '--zip_safe'
     cmd = '$TOOLS_PEX -s "%s" -m "%s" %s --interpreter_options="%s" --stamp="$RULE_HASH"' % (
         shebang, CONFIG.PYTHON_MODULE_DIR, zipsafe_flag, CONFIG.PYTHON.INTERPRETER_OPTIONS)
     if site:
         cmd += ' -S'
 
+    cmd += f'; mv __main__.py {name}_main.py'
```
[the remark this answers is not in the document]
Comment (samwestmoreland) [reply build_defs/python.build_defs:141]: The pex tool now creates `__main__.py` rather than a `.pex.zip` because we're not zipping in the dependencies now, just writing a py file

File: tools/please_pex/pex/pex.go (hunk 10 above)
Comment (tiagovtristao) [reply tools/please_pex/pex/pex.go]: Yes, it should be included now as a dep. Check my latest changes on https://github.com/thought-machine/please/pull/2213 which might point you in the right direction

File: build_defs/python.build_defs [hunk 15]
```diff
@@ -196,6 +202,21 @@ def python_binary(name:str, main:str, srcs:list=[], resources:list=[], out:str=N
     )
 
 
+def _set_shebang(interpreter:str, interpreter_options:str, site:bool):
+    """Returns a shebang to pass to arcat to inject at the top of a pex"""
+    # Assume the path to the interpreter is relative
+    shebang = f'/usr/bin/env {interpreter}'
+    if not site:
+        if ' ' in shebang:
```
Comment (Tatskaari) [inline build_defs/python.build_defs:210]: I'm not sure i understand what's going on here. We just set shebang above so this is always going to be true?

File: .plzconfig [hunk 16]
```diff
@@ -6,29 +6,38 @@ PreloadBuildDefs = test/build_defs/test.build_defs
...
+[Java]
+JarCatTool = @self//tools:arcat
 
 [PluginDefinition]
 name = python
 
 [PluginConfig "default_interpreter"]
 ConfigKey = DefaultInterpreter
 DefaultValue = python3
+Optional = true
```
Comment (Tatskaari) [inline .plzconfig:21]: Do we really want this to be optional? That means I could do 
```
[Proto "python"]
DefaultInterpreter = 
```

And `CONFIG.PYTHON.DEFAULT_INTERPRETER` would be unset. This seems like it would break things.

File: third_party/go/BUILD [hunk 17]
```diff
@@ -1,7 +1,82 @@
+go_toolchain(
```
Comment (Tatskaari) [inline third_party/go/BUILD:1]: What do you think about writing the pex tool in python at some point so we can avoid depending on go here?

File: third_party/python/BUILD [hunk 18]
```diff
@@ -59,15 +62,12 @@ python_wheel(
 python_wheel(
     name = "xmlrunner",
     package_name = "unittest_xml_reporting",
-    hashes = ["530e703bf8f90a867ce42a524fc4cc99c50c0a1522ce7bd010a783698bd4cac9"],
     version = "1.11.0",
     deps = [":six"],
 )
 
-python_wheel(
+pip_library(
```
Comment (Tatskaari) [inline third_party/python/BUILD:69]: Is there a reason you want to use `pip_library()` here instead of `python_wheel()`? `python_wheel()` is generally more reliable.

Comment (samwestmoreland) [reply third_party/python/BUILD:69]: I think this was because the wheel resolver tool depends on a few things including `six` so they can no longer be python_wheels otherwise we get cyclic dependencies.

File: third_party/go/BUILD (hunk 17 above)
Comment (samwestmoreland) [reply third_party/go/BUILD:1]: Added to my to-do list. Will have a look into it.

File: .plzconfig (hunk 16 above)
Comment (samwestmoreland) [reply .plzconfig:21]: Yeah good point! I think actually most of them shouldn't be optional. I've changed them.

File: build_defs/python.build_defs (hunk 15 above)
Comment (samwestmoreland) [reply build_defs/python.build_defs:210]: Trying to remember what was going on here, but you're right that does look weird. Think I've fixed it.

File: build_defs/python.build_defs [hunk 19]
```diff
@@ -196,6 +202,18 @@ def python_binary(name:str, main:str, srcs:list=[], resources:list=[], out:str=N
...
 
 
+def _set_shebang(interpreter:str, interpreter_options:str, site:bool):
+    """Returns a shebang to pass to arcat to inject at the top of a pex"""
+    # Assume the path to the interpreter is relative
+    shebang = f'/usr/bin/env {interpreter}'
+    if not site:
+        shebang = f'#!/bin/sh\nexec {shebang} -S \\\\$0 \\\"\\\\$@\\\"'
+    if not shebang.startswith('#!'):
+        shebang = '#!' + shebang
```
Comment (Tatskaari) [inline build_defs/python.build_defs:212]: Can't this just be
```python
if site:
    return f'#!/usr/bin/env {interpreter}'
else: 
    return f'#!/bin/sh\nexec {shebang} -S \\\\$0 \\\"\\\\$@\\\"'
```

Also, is there a reason this need to change so much? The old shebang code was just:
```
    shebang = shebang or interpreter or CONFIG.PYTHON.DEFAULT_INTERPRETER
```

I think the pex tool does something very similar to this, so maybe we just need to tell it which python interpreter we're using? Looks like it might just be adding that to the beginning of the pex main `.py` file here rather than the actual `.pex`?

File: tools/wheel_resolver/src/wheel_tags/tags.py [hunk 20]
```diff
@@ -15,16 +15,25 @@ def is_compatible(wheel, archs):
...
     # taglist is a list of tags that we've got either from the user
     # or we've auto-generated them for this system
     taglist = generate_tags_from_all_archs(archs)
+    # TODO: Could probably cache these? ^^
+
+    if taglist is None:
+        logging.critical("No tags generated")
+    # else:
+    #     len_taglist = sum(1 for _ in taglist)
+    #     logging.critical('Generated %s tags', len_taglist)
```
Comment (Tatskaari) [inline tools/wheel_resolver/src/wheel_tags/tags.py:31]: Did you mean to make these changes? Looks like it mgith be left over from another PR.

File: tools/please_pex/pex/pex.go [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+		noSite:         noSite,
+		realEntryPoint: toPythonPath(entryPoint),
+		pexStamp:       stamp,
+	}
+	pw.SetShebang(interpreter, options)
+	return pw
+}
+
+// SetShebang sets the leading shebang that will be written to the file.
+func (pw *Writer) SetShebang(shebang string, options string) {
```
Comment (Tatskaari) [inline tools/please_pex/pex/pex.go:59]: Ah I see, I don't think this is used anymore? Maybe we should just delete this.

**Review by Tatskaari — APPROVED** [review]
Looks good! Not sure I fully understand how the pex main is working but it doesn't look like it's changed much. Was a little confused as to how the shebangs work, but I think i've figured it out. 

Happy for it to land after a little cleanup.

File: build_defs/python.build_defs (hunk 19 above)
Comment (samwestmoreland) [reply build_defs/python.build_defs:212]: Yeah it could definitely be the top one. I can't remember exactly but I think the idea behind changing this was that, like you said below, we're not doing any shebang logic in the pex tool anymore and we want to do it here instead.
